### PR TITLE
Fixed whileLoading is not rendered while loading

### PR DESCRIPTION
**Pull request by LBrian**
PR for #2264 

We agreed earlier to use React `Suspense` to solve this issue, however deep dive into the code base, the way `PageLoader` was implemented (concept) is pretty much similar to `Suspense`, with that said using `Suspense` will become breaking change as we no longer be able to pass `loading` state down to children (`usePageLoadingContext` becomes deprecated). Thus, I believe this is a neat and easy fix without introducing a breaking change.

### CODE REVIEW

**Comment by peterp** [comment]
I'm all for incremental adoption!

**Review by Tobbe — APPROVED** [review]

**Comment by dac09** [comment]
This is looking really good to me, just have to make sure no flickering gets introduced accidentally! I'll  do a quick check of this and then merge.

**Review by dac09 — CHANGES_REQUESTED** [review]
Hey @LBrian thanks again for this. Just tried this out, and works quite well, no unnecessary flickers/renders. However there's a mismatch between what `whileLoading` means. 

Tried to explain in this tape here: https://s.tape.sh/1DAenGaT
(i'll attempt to explain in this comment too 🙈)

- when I use `whileLoading` on private page, it renders _only_ the whileLoading component, and none of my layouts/page
- with this change, as the page chunk is loading, it already has the layout/page rendered but then will render the whileLoading component _inside_ the page

In essence, these should not be the same component - because one component is `whileLoadingAuthStatus` and the other is `whileLoadingPageChunk`. I'll get back to you on the naming of the props shortly, unless you can think of a better solution

**Comment by LBrian** [comment]
> Hey @LBrian thanks again for this. Just tried this out, and works quite well, no unnecessary flickers/renders. However there's a mismatch between what `whileLoading` means.
> 
> Tried to explain in this tape here: https://s.tape.sh/1DAenGaT
> (i'll attempt to explain in this comment too 🙈)
> 
> * when I use `whileLoading` on private page, it renders _only_ the whileLoading component, and none of my layouts/page as its retrieving my auth status. So the whole page is the whileLoading component
> * with this change, as the page chunk is loading, it already has the layout/page rendered but then will render the whileLoading component _inside_ the page, so I think its serving a slightly different purpose
> 
> Relevant snippet of my Routes, where I wrap the page in a layout with Set
> 
> ```
>  <Router pageLoadingDelay={500}>
>       <Private unauthenticated="login">
>         <Set wrap={DashboardLayout} prerender>
>           <Route
>             path="/dashboard"
>             page={DashboardHomePage}
>             name="dashboard"
>             whileLoading={DashboardLoader} //DashboardLoader is DashboardLayout+Sheen bars
>           />
> ```
> 
> In essence, these should not be the same component - because one component is `whileLoadingAuthStatus` and the other is `whileLoadingPageChunk`. I'll get back to you on the naming of the props shortly, unless you can think of a better solution

@dac09 Thanks for verifying the PR and the video does save some time so we can pinpoint the issue quickly, since Im a big fan for screen recording for communication with team members. (btw, love your Tape 😄 ) 

This is similar to what I suggested at the beginning [here](https://github.com/redwoodjs/redwood/issues/2264#issuecomment-862857490), and the reason why I minimised changes  in this PR coz we didn't seem to come down to a conclusion for introducing a new prop on `Set` and `Private` for authentication purpose.

Basically, `Set` shouldn't be picked up `whileLoading` for authoring transition at all, `whileLoading` should only be used for page chunks loading, with that said, we should do
1. introduce a new prop called `whileAuthenticating` (can be discussed) to `Set` (includes `Private`) for auth transition only
2. leave `whileLoading` the same for `PageLoader` (in this PR)

Thus, for **your** case, you can then set `whileAuthenticating={DashboardLoader}` to `Private` for authentication transition and set `<Route whileLoading={PageChunksLoader} />` for loading chunks, hopefully this makes sense for you.

**Comment by LBrian** [comment]
@dac09 Just did a quick change for context, lmk what you think, naming can be changed

https://github.com/redwoodjs/redwood/pull/2889/commits/8c21e142fca15d43522594c31cabd632b717ca4d

**Comment by dac09** [comment]
This looks great @LBrian thanks for the update. I'll just confirm on the naming and we should be good to go.

**Comment by dac09** [comment]
Hey @LBrian - decision made on naming. Can we renaming the props to be `whileLoadingAuth` and `whileLoadingPage` please?

Thanks a bunch!

**Comment by Tobbe** [comment]
Should we keep `whileLoading` for a while to be backwards compatible?

The current solution to the `whileLoadingPage` scenario is `const { loading } = usePageLoadingContext()`. Should we keep that as well? Or remove it to keep down the complexity?

**Comment by dac09** [comment]
I think it's better that we make the breaking changes now, its quite a localised change too.

**Comment by dac09** [comment]
> The current solution to the whileLoadingPage scenario is const { loading } = usePageLoadingContext(). Should we keep that as well? Or remove it to keep down the complexity?

I'm going to suggest we keep that functionality there - not sure if there's other use cases for it, but docs should definitely be updated!

**Comment by LBrian** [comment]
@dac09 You are way faster than me 😄 I was thinking to update document as well before updating here, anyway I will create a PR for document later today!

**Comment by peterp** [comment]
For what it's worth if we're going to introduce a breaking change then I think we should just go the `Suspense` route?

**Comment by LBrian** [comment]
I am not in the core team so I have zero knowledge about how the team works and the roadmap, I would suggest to roll this out first in the coming minor version release to correct the misleading concept of `whileLoading` since this small breaking is simply just rename the prop from `whileLoading` to `whileLoadingPage`.

`Suspense` for me is a bigger scale of breaking change compare with this one, so still good to iterate it, that's just my opinion and again I will leave this to the core team to decide it.

**Comment by dac09** [comment]
Looking good @LBrian - thanks for this. 💯 

I'll make it a non breaking change later if there's any objections from the team, but merging now.


For future reference, the way I'm testing this is:
1. `rwt link` after pulling this branch
2. `yarn rw build && yarn rw serve`
3. Slow everything down on Chrome, and enjoy the loader magic

**Comment by dac09** [comment]
Actually... on my second pass... it looks like its changed the behaviour on prerender. I can't see anywhere obvious in the code that this might be happening.. apart from the flattenArray bit in Set

What I expect: Should render the page loader for private pages
What I see: blank page - so it thinks whileLoadingAuth

**Review by dac09 — APPROVED** [review]
